### PR TITLE
[composable-controller] Narrow class and messenger types by parameterizing over child controllers list

### LINKED ISSUE #3627: [composable-controller] Make class and messenger generic upon child controllers
## Explanation

Currently, the allow lists of `ComposableControllerMessenger` are set to `string`, which is too permissive. Each `ComposableController` instance needs typing and allow lists that are specific to its set of input child controllers.

However, there's no simple way to achieve this using `BaseController`, as controller-messenger allow lists are defined for the controller class singleton, not controller instances.

Instead, `ComposableController` and `ComposableControllerMessenger` need to be made polymorphic upon the `controllers` constructor property, which represents the list of child controllers to be composed. 

## Requirements

- A generic parameter `ControllerList` (or maybe `ChildControllers`) needs to be added to the `ComposableController` class and `ComposableControllerMessenger` type.
- `ControllerList` should be constrained, but not defined by the least supertype for all base-controller instances (currently `ControllerInstance`).
- ~`ControllerList` should be used to derive the return type of `flatState()` at the type level.~
- `ControllerList` should be used to derive the state type for `ComposableController` instances.
- `ControllerList` should be used to derive the allow lists of `ComposableControllerMessenger`. 
  - The allow lists should at minimum include the `stateChange` events for all child controllers that are subscribed to in the `#updateChildController()` method. 

## References

- See https://github.com/MetaMask/core/pull/3590#discussion_r1414460107

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

Currently, the allow lists of `ComposableControllerMessenger` are set to `string`, which is too permissive. Each `ComposableController` instance needs typing and allow lists that are specific to its set of input child controllers.

To achieve this, the `ComposableController` class and its messenger are made polymorphic upon the `ControllerState` type, which defines the shape of the composed state.

## References

- Closes #3627

### Added

- Adds and exports new types: ([#3952](https://github.com/MetaMask/core/pull/3952))
  - `RestrictedControllerMessengerConstraint`, which is the narrowest supertype of all controller-messenger instances.
  - `LegacyControllerStateConstraint`, a universal supertype for the controller state object, encompassing both BaseControllerV1 and BaseControllerV2 state.
  - `ComposableControllerStateConstraint`, the narrowest supertype for the composable controller state object.

### Changed

- **BREAKING:** The `ComposableController` class is now a generic class that expects one generic argument `ComposableControllerState` ([#3952](https://github.com/MetaMask/core/pull/3952)).
  - **BREAKING:** For the `ComposableController` class to be typed correctly, any of its child controllers that extend `BaseControllerV1` must have an overridden `name` property that is defined using the `as const` assertion.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by MajorLift** [comment]
@Gudahtt The only user-facing generic parameter in `ComposableController` is `ChildControllers`. The other five parameters are automatically derived from `ChildControllers` without any further user input, using the expressions in their generic parameter defaults. 

These "optional" generic parameters are essentially internal variables of a "ComposableController" type-level function that are evaluated when the function is called with a "ChildControllers" argument. 

This gives us exactly accurate types for `ComposedControllerState` and `ComposedControllerMessenger` that (as "scoped/internal variables") are accessible from within each composable-controller instance. Using types that are wider than the current generic params would be imo an unnecessary regression in terms of type safety.

I'll try to find a way in the jsdoc to make it immediately clear that `ComposableController` exposes only one generic parameter. I'm thinking only leaving `ChildControllers` in the jsdoc, and moving information about the "private" variables to in-line comments should work.

**Comment by Gudahtt** [comment]
The prototype branch I linked also has exactly accurate types. I was not suggesting using wider types, just deriving things the other way (deriving controller types from state, rather than state from controller). It seems simpler to me to have fewer generic type parameters, even if they are optional.

**Comment by MajorLift** [comment]
- The `controllers` constructor option not being constrained shouldn't be an issue if the `controllers` input is used as the source of truth at runtime. Even if a generic param `ChildControllers` is assigned to `controllers`, the generic param will be inferred from `controllers` if left omitted.

- The biggest difference between our approaches seems to be the types deriving from `ChildControllers` vs. `ComposableControllerState`. Needing to supply the state object everywhere might make the API a bit awkward (example below).

But if the composed state object is expected to be readily available (without needing to be derived from `ChildControllers` or something else), I think either way should work.

```ts
const controllerMessenger = new ControllerMessenger<
        never,
        | ComposableControllerEvents<{
            FooController: { foo: 'foo' };
            QuzController: { quz: 'quz' };
          }>
        | FooControllerEvent
        | QuzControllerEvent
      >();
```

- One clear advantage of the prototype branch is that it lets us export controller instance-specific types. If this is important, we should definitely go with this approach.

File: packages/composable-controller/CHANGELOG.md [hunk 1]
```diff
@@ -25,10 +25,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 
 - **BREAKING:** Passing a non-controller into `controllers` constructor option now throws an error ([#3904](https://github.com/MetaMask/core/pull/3904))
-- **BREAKING:** The `AllowedAction` parameter of the `ComposableControllerMessenger` type is narrowed from `string` to `never`, as `ComposableController` does not use any external controller actions ([#3904](https://github.com/MetaMask/core/pull/3904))
+- **BREAKING:** The `AllowedActions` parameter of the `ComposableControllerMessenger` type is narrowed from `string` to `never`, as `ComposableController` does not use any external controller actions. ([#3904](https://github.com/MetaMask/core/pull/3904))
```
Comment (mcmire) [inline packages/composable-controller/CHANGELOG.md:28]: It seems that we are updating the changelog for an older version, is this intentional? Are there changes we forgot to include for this version?

File: packages/composable-controller/src/ComposableController.ts [hunk 2]
```diff
@@ -99,43 +107,108 @@ export function isBaseController(
...
 
+type GetChildControllers<
+  ComposableControllerState,
+  P extends keyof ComposableControllerState = keyof ComposableControllerState,
+> = P extends string
+  ? ComposableControllerState[P] extends StateConstraint
+    ? { name: P; state: ComposableControllerState[P] }
+    : BaseControllerV1<
+        BaseConfig & Record<string, unknown>,
+        BaseState & ComposableControllerState[P]
```
Comment (mcmire) [inline packages/composable-controller/src/ComposableController.ts:197]: Nit: Can we use a non-one-letter name here for clarify? Perhaps `ControllerName`?

``` suggestion
  ControllerName extends keyof ComposableControllerState = keyof ComposableControllerState,
> = Key extends string
  ? ComposableControllerState[ControllerName] extends StateConstraint
    ? { name: ControllerName; state: ComposableControllerState[ControllerName] }
    : BaseControllerV1<
        BaseConfig & Record<string, unknown>,
        BaseState & ComposableControllerState[ControllerName]
```

**Review by mcmire — COMMENTED** [review]
Had some questions, but overall makes sense.

File: packages/composable-controller/CHANGELOG.md (hunk 1 above)
Comment (MajorLift) [reply packages/composable-controller/CHANGELOG.md]: Fixed to align with final changelog in PR description: https://github.com/MetaMask/core/pull/3952/commits/365aa0214cce99a9065e2901738ac3116b2619a5

File: packages/composable-controller/src/ComposableController.ts (hunk 2 above)
Comment (MajorLift) [reply packages/composable-controller/src/ComposableController.ts]: Fixed here: https://github.com/MetaMask/core/pull/3952/commits/25f2f0b6574dc5dda7cc7832cbc5a8833f240e75

**Review by mcmire — APPROVED** [review]
Looks good!

**Pull request merged**